File: 2010/S5.cpp
#include <bits/stdc++.h>

using namespace std;

class Node {
public:
    Node *left, *right;
    int value;
    vector<int> nutrients;

    Node(Node *left, Node *right, int value = 0) : left(left), right(right), value(value) {}
};

string trim(const string& str) {
    size_t first = str.find_first_not_of(" \n\r\t");
    size_t last = str.find_last_not_of(" \n\r\t");

    if (first == string::npos || last == string::npos)
        return "";
    
    return str.substr(first, (last - first + 1));
}

Node* createNode(string s, int x) {
    s = trim(s);
    if (s[0] != '(') {
        return new Node(nullptr, nullptr, stoi(s));
    } else {
        string inner = trim(s.substr(1, s.size() - 2));
        int i = 0, count = 0;
        if (inner[0] == '(') {
            count = 1;
            i = 1;
            while (count > 0) {
                if (inner[i] == '(') count++;
                else if (inner[i] == ')') count--;
                i++;
            }
        } else {
            i = inner.find(' ');
        }
        Node *left = createNode(inner.substr(0, i), x);
        Node *right = createNode(inner.substr(i + 1), x);
        return new Node(left, right);
    }
}

void solve(Node* node, int x) {
    node->nutrients.resize(x + 1);
    if (node->left == nullptr && node->right == nullptr) {
        for (int i = 0; i <= x; ++i) {
            node->nutrients[i] = node->value + i;
        }
    } else {
        solve(node->left, x);
        vector<int> left(x + 1, 0);
        for (int i = 0; i <= x; ++i) {
            int maxLeft = 0;
            for (int j = 0; j <= i; ++j) {
                maxLeft = max(maxLeft, min((1 + j) * (1 + j), node->left->nutrients[i - j]));
            }
            left[i] = maxLeft;
        }

        solve(node->right, x);
        vector<int> right(x + 1, 0);
        for (int i = 0; i <= x; ++i) {
            int maxRight = 0;
            for (int j = 0; j <= i; ++j) {
                maxRight = max(maxRight, min((1 + j) * (1 + j), node->right->nutrients[i - j]));
            }
            right[i] = maxRight;
        }

        for (int i = 0; i <= x; ++i) {
            int maxCombined = 0;
            for (int j = 0; j <= i; ++j) {
                maxCombined = max(maxCombined, left[j] + right[i - j]);
            }
            node->nutrients[i] = maxCombined;
        }
    }
}

int main() {
    string line;
    getline(cin, line); // Read the tree representation
    int x;
    cin >> x;

    Node* root = createNode(line, x);
    solve(root, x);
    cout << root->nutrients[x] << endl;

    return 0;
}

File: 2010/S5.py
import sys
input = sys.stdin.readline

string = input().strip()
x = int(input())

class Node:
    def __init__(self, left, right):
        self.left = left
        self.right = right
        self.nutrients = [0] * (x + 1)
class Leaf:
    def __init__(self, value):
        self.value = value
        self.nutrients = [0] * (x + 1)

def createNode(s):
    if not s.startswith('('):
        return Leaf(int(s))
    else:
        s = s[1:len(s) - 1].strip()
        i = 0
        if s.startswith('('):
            count = 1
            i = 1
            while count > 0:
                if s[i] == '(':
                    count += 1
                elif s[i] == ')':
                    count -= 1
                i += 1
        else:
            i = s.index(' ')
        return Node(createNode(s[:i]), createNode(s[i + 1:]))

def solve(node, x):
    if isinstance(node, Leaf): # leaf node
        for i in range(x + 1):
            node.nutrients[i] = node.value + i # fill up nutrients[]
    else:
        solve(node.left, x) # fill up left subtree
        left = [0] * (x + 1) # max of left subtree using i agents
        for i in range(x + 1): # checking max from agents to branches
            maxLeft = 0
            for j in range(i + 1):
                maxLeft = max(maxLeft, min((1 + j) ** 2, node.left.nutrients[i - j])) # used j agents on branches, so i - j left on leaves
            left[i] = maxLeft

        solve(node.right, x) # fill up right subtree
        right = [0] * (x + 1) # max of right subtree using i agents
        for i in range(x + 1): # checking max from agents to branches
            maxRight = 0
            for j in range(i + 1):
                maxRight = max(maxRight, min((1 + j) ** 2, node.right.nutrients[i - j])) # used j agents on branches, so i - j left on leaves
            right[i] = maxRight
        
        for i in range(x + 1): # combining left and right subtrees
            maxCombined = 0
            for j in range(i + 1):
                maxCombined = max(maxCombined, left[j] + right[i - j]) # used j agents on left subtree, so i - j left on right
            node.nutrients[i] = maxCombined

root = createNode(string)
solve(root, x)
print(root.nutrients[x])